### PR TITLE
Feature/rulemapping

### PR DESCRIPTION
**Pull request by brysonjbest**
- [x] Added rule mapping feature that generates input/output/results schemas from rules. 
- [x] Added testing suite for rule mapping.
Schema generation works:
- [x] on files
- [x] post requests with rule data
- [x] post requests with trace data from run rules

### CODE REVIEW

File: src/api/ruleMapping/ruleMapping.service.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        }));
+      } else {
+        return (node.content?.outputs || []).map((outputField: { id: any; name: any; type: any; field: any }) => ({
+          id: outputField.id,
+          name: outputField.name,
+          type: outputField.type,
+          property: outputField.field,
+        }));
+      }
+    });
```
Comment (timwekkenbc) [inline src/api/ruleMapping/ruleMapping.service.ts:44]: Are extractInputs and extractOutputs essentially the same? Maybe unify them under the same function or at least have a function they both call so code isn't repeated. Also, make use of your types from your interface (InputField/OutputField)

File: src/api/ruleMapping/ruleMapping.interface.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      [key: string]: any;
+    } | null;
+    performance?: string;
+    traceData?: {
+      index: number;
+      reference_map: Record<string, any>;
+      rule: Record<string, any>;
+    };
+  };
+}
```
Comment (timwekkenbc) [inline src/api/ruleMapping/ruleMapping.interface.ts:56]: For types like this one, is it possible to get them from the zen-engine library itself? Like this one seems to essentially be an extension of `ZenEngineTrace`

File: src/api/ruleMapping/ruleMapping.interface.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  type: string;
+  field: string;
+}
+
+export interface OutputField {
+  id: string;
+  name: string;
+  type: string;
+  field: string;
+}
```
[the remark this answers is not in the document]
Comment (brysonjbest) [reply src/api/ruleMapping/ruleMapping.interface.ts]: updated: [a18ab51](https://github.com/bcgov/brms-api/pull/10/commits/a18ab51b8b779184cd2e4784894af4d30f4716b6)

File: src/api/ruleMapping/dto/evaluate-rulemapping.dto.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+export class EvaluateRuleMappingDto {
+  @IsArray()
+  @ValidateNested()
+  @Type(() => Node)
+  nodes: Node[];
+  edges: Edge[];
+}
+export class EvaluateRuleRunSchemaDto {
+  @ValidateNested()
+  trace: TraceObject;
```
[the remark this answers is not in the document]
Comment (brysonjbest) [reply src/api/ruleMapping/dto/evaluate-rulemapping.dto.ts]: updated: [a18ab51](https://github.com/bcgov/brms-api/pull/10/commits/a18ab51b8b779184cd2e4784894af4d30f4716b6)

File: src/api/ruleMapping/ruleMapping.service.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+          Object.assign(input, traceData.input);
+        }
+      }
+    }
+    return { input, output };
+  }
+
+  // generate a rule schema from a given local file
+  async ruleSchemaFile(filePath: string): Promise<any> {
+    const documentsService = new DocumentsService();
```
[the remark this answers is not in the document]
Comment (brysonjbest) [reply src/api/ruleMapping/ruleMapping.service.ts]: updated: [beebed2](https://github.com/bcgov/brms-api/pull/10/commits/beebed291d006140a53583a97e310207c242ce90)

**Comment by brysonjbest** [comment]
Updated to incorporate latest brms rules folder location changes on dev.

File: src/api/ruleMapping/ruleMapping.interface.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (timwekkenbc) [reply src/api/ruleMapping/ruleMapping.interface.ts]: Why the change to using classes instead of interfaces?

**Review by timwekkenbc — APPROVED** [review]
Looks good to me!

**Pull request merged**